text/0111-auto-resource-detection.md

```diff
@@ -0,0 +1,110 @@
+
+Note this means that if the user wants to make use of auto-detected resource information, they will be required to explicitly pass the resource to the trace and/or metric provider(s).
+
+If multiple detectors are configured, and more than one of these successfully detects a resource, the resources will be merged according to the Merge interface already defined in the specification, i.e. the earliest matched resource's attributes will take precedence. Each detector may be run in parallel, but to ensure deterministic results, the resources must be merged in the order the detectors were added.
+
+A default implementation of a detector that reads resource data from the `OTEL_RESOURCE` environment variable will be included in the SDK. The environment variable will contain of a list of key value pairs, and these are expected to be represented in a format similar to the [W3C Correlation-Context](https://github.com/w3c/correlation-context/blob/master/correlation_context/HTTP_HEADER_FORMAT.md#header-value), i.e.: `key1=value1,key2=value2`. This detector must always be configured as the first detector by default.
+
+Custom resource detectors related to specific environments (e.g. specific cloud vendors) must be implemented outside of the SDK, and users will need to import these separately.
+
+In order to allow users to be able to use different detectors for different trace or metric providers, the detector(s) must be configured against a resource provider, and users can create multiple providers if desired.
+
+## Internal details
+
+As described above, the following will be added to the Resource SDK specification:
+
+- An interface for "detectors", to retrieve resource information
+- Specification for the resource provider, which can have custom detectors added, and will return a merged resource
+- Details of the default "From Environment Variable" detector implementation described above
+
+This is a relatively small proposal so is easiest to explain the details with a code example:
+
+### Usage
+
+The following example in Go creates a trace provider that uses resource information automatically detected from AWS or GCP:
+
+Assumes a dependency has been added on the `otel api`, `otel sdk`, `otel awsdetector`, and `otel gcpdetector` packages.
+
+```go
+rp := resource.NewMustProvider()
+rp.AddDetectors(awsdetector.EC2, gcpdetector.GCE, gcpdetector.GKE)
+resource := rp.AutoDetect(ctx)
+tp := sdktrace.NewProvider(sdktrace.WithResource(resource))
+```
+
+There is no change to how the resource is given to the provider, so no changes are required to the existing spec / implementation.
+
+### Components
+
+#### Detector
+
+The Detector interface will simply return a Resource:
+
+```go
+type Detector interface {
+    Detect(ctx context.Context) (*Resource, error)
+}
+```
+
+If a detector is not able to detect a resource, it must return an empty resource so that the result of each call to `Detect` can be merged.
```

Comment:
`nil` resource should work fine for Go as well.
In other languages it might be different, as long as it leads to the same result.

I think there is also a difference between empty and missing resource in OTLP? See https://github.com/open-telemetry/opentelemetry-proto/blob/b54688569186e0b862bf7462a983ccf2c50c0547/opentelemetry/proto/metrics/v1/metrics.proto#L30

Reply:
Hmm need to confirm if that is an important distinction. Note the current typescript implemention returns an empty resource.